In [3]:
!pip install spacy

  Obtaining dependency information for spacy from https://files.pythonhosted.org/packages/ca/f3/609bb7512cad1f02af13daa23aa433b931da34c502211f29fd47dceff624/spacy-3.7.2-cp311-cp311-macosx_11_0_arm64.whl.metadata
  Using cached spacy_legacy-3.0.12-py2.py3-none-any.whl (29 kB)
  Obtaining dependency information for spacy-loggers<2.0.0,>=1.0.0 from https://files.pythonhosted.org/packages/33/78/d1a1a026ef3af911159398c939b1509d5c36fe524c7b644f34a5146c4e16/spacy_loggers-1.0.5-py3-none-any.whl.metadata
  Obtaining dependency information for murmurhash<1.1.0,>=0.28.0 from https://files.pythonhosted.org/packages/7a/05/4a3b5c3043c6d84c00bf0f574d326660702b1c10174fe6b44cef3c3dff08/murmurhash-1.0.10-cp311-cp311-macosx_11_0_arm64.whl.metadata
  Obtaining dependency information for cymem<2.1.0,>=2.0.2 from https://files.pythonhosted.org/packages/d7/f6/67babf1439cdd6d46e4e805616bee84981305c80e562320c293712f54034/cymem-2.0.8-cp311-cp311-macosx_11_0_arm64.whl.metadata
  Obtaining dependency information 

In [4]:
!python -m spacy download en_core_web_lg

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 587.7/587.7 MB 3.6 MB/s eta 0:00:0000:0100:02
✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_lg')


# Importing Data

In [5]:
import pandas as pd

In [6]:


data = pd.read_csv('/Users/vijaykumardevalla/Documents/Unstructured Data Analytics/Assignment 2/restaurant_reviews_az.csv')

In [7]:
data.head()

review_id                 user_id             business_id  \
0  IVS7do_HBzroiCiymNdxDg  fdFgZQQYQJeEAshH4lxSfQ  sGy67CpJctjeCWClWqonjA   
1  QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2  oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3  E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
4  Rd222CrrnXkXukR2iWj69g  LPxuausjvDN88uPr-Q4cQA  CA5BOxKRDPGJgdUQ8OUOpw   

   stars  useful  funny  cool  \
0      3       1      1     0   
1      5       1      1     1   
2      5       1      0     0   
3      5       0      0     0   
4      4       1      0     0   

                                                text             date  \
0  OK, the hype about having Hatch chili in your ...  1/27/2020 22:59   
1  Pandemic pit stop to have an ice cream.... onl...   4/19/2020 5:33   
2  I was lucky enough to go to the soft opening a...  2/29/2020 19:43   
3  I've gone to claim Jumpers all over the US and...  3/14/2020 21:47   
4  If you haven't been  to Maynard's kitchen, it'...  1/17/2020 20:32   

   Sentiment  
0          1  
1          1  
2          1  
3          1  
4          1

In [8]:
import spacy #load spacy
nlp = spacy.load("en_core_web_lg")
def normalize(review, lowercase, remove_stopwords):
    if lowercase:
        review = review.lower()
    doc = nlp(review)
    lemmatized = list()
    for token in doc:
        if not remove_stopwords or (remove_stopwords and not token.is_stop):
            lemmatized.append(token.lemma_)
    return " ".join(lemmatized)
data['processed'] = data['text'].apply(normalize, lowercase=True, remove_stopwords=True)

In [9]:
data['Sentiment'].value_counts()

Sentiment
1    33268
0    14879
Name: count, dtype: int64

In [11]:
!pip install vaderSentiment

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 126.0/126.0 kB 6.4 MB/s eta 0:00:00


In [12]:
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

In [13]:
sentiment = SentimentIntensityAnalyzer()



In [16]:
# Define a function that applies sentiment analysis to a text
def analyze_sentiment(text):
    return sentiment.polarity_scores(text)

# Apply the function to each row in the 'text' column
data['sentiment'] = data['processed'].apply(analyze_sentiment)


In [17]:
data.head()

review_id                 user_id             business_id  \
0  IVS7do_HBzroiCiymNdxDg  fdFgZQQYQJeEAshH4lxSfQ  sGy67CpJctjeCWClWqonjA   
1  QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2  oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3  E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
4  Rd222CrrnXkXukR2iWj69g  LPxuausjvDN88uPr-Q4cQA  CA5BOxKRDPGJgdUQ8OUOpw   

   stars  useful  funny  cool  \
0      3       1      1     0   
1      5       1      1     1   
2      5       1      0     0   
3      5       0      0     0   
4      4       1      0     0   

                                                text             date  \
0  OK, the hype about having Hatch chili in your ...  1/27/2020 22:59   
1  Pandemic pit stop to have an ice cream.... onl...   4/19/2020 5:33   
2  I was lucky enough to go to the soft opening a...  2/29/2020 19:43   
3  I've gone to claim Jumpers all over the US and...  3/14/2020 21:47   
4  If you haven't been  to Maynard's kitchen, it'...  1/17/2020 20:32   

   Sentiment                                          processed  \
0          1  ok , hype have hatch chili burger overrate . o...   
1          1  pandemic pit stop ice cream .... plain sundae ...   
2          1  lucky soft opening let tell ... good . beer wi...   
3          1  go claim jumper disappoint location different ...   
4          1    maynard kitchen , time ! hope dinner , sure ...   

                                           sentiment  
0  {'neg': 0.02, 'neu': 0.652, 'pos': 0.328, 'com...  
1  {'neg': 0.169, 'neu': 0.632, 'pos': 0.199, 'co...  
2  {'neg': 0.0, 'neu': 0.613, 'pos': 0.387, 'comp...  
3  {'neg': 0.069, 'neu': 0.587, 'pos': 0.344, 'co...  
4  {'neg': 0.048, 'neu': 0.67, 'pos': 0.282, 'com...

In [37]:
#Splitting the data into trainig and testing
from sklearn.model_selection import train_test_split
X_train, X_test, Y_train, Y_test = train_test_split(data['processed'], data['Sentiment'], test_size=0.2, random_state=5)


In [38]:
data.head()

review_id                 user_id             business_id  \
0  IVS7do_HBzroiCiymNdxDg  fdFgZQQYQJeEAshH4lxSfQ  sGy67CpJctjeCWClWqonjA   
1  QP2pSzSqpJTMWOCuUuyXkQ  JBLWSXBTKFvJYYiM-FnCOQ  3w7NRntdQ9h0KwDsksIt5Q   
2  oK0cGYStgDOusZKz9B1qug  2_9fKnXChUjC5xArfF8BLg  OMnPtRGmbY8qH_wIILfYKA   
3  E_ABvFCNVLbfOgRg3Pv1KQ  9MExTQ76GSKhxSWnTS901g  V9XlikTxq0My4gE8LULsjw   
4  Rd222CrrnXkXukR2iWj69g  LPxuausjvDN88uPr-Q4cQA  CA5BOxKRDPGJgdUQ8OUOpw   

   stars  useful  funny  cool  \
0      3       1      1     0   
1      5       1      1     1   
2      5       1      0     0   
3      5       0      0     0   
4      4       1      0     0   

                                                text             date  \
0  OK, the hype about having Hatch chili in your ...  1/27/2020 22:59   
1  Pandemic pit stop to have an ice cream.... onl...   4/19/2020 5:33   
2  I was lucky enough to go to the soft opening a...  2/29/2020 19:43   
3  I've gone to claim Jumpers all over the US and...  3/14/2020 21:47   
4  If you haven't been  to Maynard's kitchen, it'...  1/17/2020 20:32   

   Sentiment                                          processed  \
0          1  ok , hype have hatch chili burger overrate . o...   
1          1  pandemic pit stop ice cream .... plain sundae ...   
2          1  lucky soft opening let tell ... good . beer wi...   
3          1  go claim jumper disappoint location different ...   
4          1    maynard kitchen , time ! hope dinner , sure ...   

                                           sentiment  
0  {'neg': 0.02, 'neu': 0.652, 'pos': 0.328, 'com...  
1  {'neg': 0.169, 'neu': 0.632, 'pos': 0.199, 'co...  
2  {'neg': 0.0, 'neu': 0.613, 'pos': 0.387, 'comp...  
3  {'neg': 0.069, 'neu': 0.587, 'pos': 0.344, 'co...  
4  {'neg': 0.048, 'neu': 0.67, 'pos': 0.282, 'com...

In [39]:
#Pre-Prcoessing and Bag of Word Vectorization using Count Vectorizer
from sklearn.feature_extraction.text import CountVectorizer
from nltk.tokenize import RegexpTokenizer
token = RegexpTokenizer(r'[a-zA-Z]+')
cv = CountVectorizer(stop_words='english',ngram_range = (1,1),tokenizer = token.tokenize, max_features = 1000)
#cv = CountVectorizer(stop_words='english',ngram_range = (1,1),tokenizer = token.tokenize, max_features = 3000, binary = True)
X_train_vect = cv.fit_transform(X_train)
X_train_vect.shape



/Users/vijaykumardevalla/anaconda3/lib/python3.11/site-packages/sklearn/feature_extraction/text.py:525: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


(38517, 1000)

In [40]:
df = pd.DataFrame(X_train_vect.toarray(), columns=cv.get_feature_names_out())
df.head()

able  absolute  absolutely  accommodate  act  actual  actually  add  \
0     0         0           0            0    0       0         0    0   
1     0         0           0            0    0       0         0    0   
2     0         0           0            0    0       0         0    0   
3     0         0           0            0    0       0         1    0   
4     0         0           0            0    0       0         0    0   

   addition  additional  ...  write  wrong  year  yelp  yes  yesterday  young  \
0         0           0  ...      0      0     0     0    0          0      0   
1         0           0  ...      0      0     1     0    1          0      0   
2         0           0  ...      0      0     0     1    0          0      0   
3         0           0  ...      0      1     0     0    0          0      0   
4         0           0  ...      0      0     0     0    0          0      0   

   yum  yummy  zero  
0    0      0     0  
1    0      0     0  
2    0      0     0  
3    0      0     0  
4    0      0     0  

[5 rows x 1000 columns]

In [41]:
# variables to store term statistics
num_of_comments = 0
unique_word = set() # using the set-type variable since it does not allow duplicates > able to count the number of unique words
num_of_token_per_comment = [] # using the list-type varailbe since we want to measure corpus-level statistics (e.g., avg, max, min, median, etc.)
num_of_token_per_comment_without_stop_words = []
total_number_of_tokens = 0 # in a corpus
unique_user = set() # using the set-type variable since it does not allow duplicates > able to count the number of unique users
date_list = [] # able to measure the number of comments by day, week, etc.
vote_count = 0
unique_submission = set() # using the set-type variable since it does not allow duplicates > able to count the number of unique submissions

In [42]:
for index, row in data.iterrows():
    text = row["processed"]
    doc = nlp(text)
    num_of_comments += 1

    # statistics regarding words
    num_of_tokens = len(doc)
    total_number_of_tokens += num_of_tokens
    token_count_without_stop_words = 0

    for token in doc:
        if token.is_stop is True:
            pass
        else:
            unique_word.add(str(token).lower())
            token_count_without_stop_words += 1

    num_of_token_per_comment.append(num_of_tokens)
    num_of_token_per_comment_without_stop_words.append(token_count_without_stop_words)

    """
    # statistics regarding users
    user_id = row["user_id"]
    unique_user.add(author_name.lower())

    # statistics regarding date
    date = row["date"]
    date_list.append(date)

    # statistics regarding reviews
    review_id = row["review_id"]
    unique_submission.add(review_id)
    """

In [43]:
#Training the model
from sklearn.naive_bayes import MultinomialNB
MNB = MultinomialNB()
MNB.fit(X_train_vect, Y_train)

MultinomialNB()

In [47]:
X_test_vect= cv.transform(X_test)
X_test_vect.shape

(9630, 1000)

In [48]:
#Evaluate the performance of the model
from sklearn import metrics
predicted = MNB.predict(X_test_vect)
performance = metrics.classification_report(Y_test,predicted, target_names= ['0', '1'])
print(performance)

              precision    recall  f1-score   support

           0       0.81      0.80      0.81      2938
           1       0.91      0.92      0.92      6692

    accuracy                           0.88      9630
   macro avg       0.86      0.86      0.86      9630
weighted avg       0.88      0.88      0.88      9630



In [49]:
from sklearn import svm
clf = svm.SVC()
clf.fit(X_train_vect, Y_train)

SVC()

In [50]:
from sklearn import metrics
predicted = clf.predict(X_test_vect)
performance = metrics.classification_report(Y_test,predicted, target_names= ['0', '1'])
print(performance)

              precision    recall  f1-score   support

           0       0.85      0.84      0.85      2938
           1       0.93      0.94      0.93      6692

    accuracy                           0.91      9630
   macro avg       0.89      0.89      0.89      9630
weighted avg       0.91      0.91      0.91      9630



In [51]:
from sklearn.feature_extraction.text import TfidfVectorizer
from nltk.tokenize import RegexpTokenizer
token = RegexpTokenizer(r'[a-zA-Z]+')
vectorizer = TfidfVectorizer(stop_words='english',ngram_range = (1,1),tokenizer = token.tokenize, max_features = 800)
X_train_tfidf = vectorizer.fit_transform(X_train)
X_train_tfidf.shape


/Users/vijaykumardevalla/anaconda3/lib/python3.11/site-packages/sklearn/feature_extraction/text.py:525: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(


(38517, 800)

In [52]:
df = pd.DataFrame(X_train_tfidf.toarray(), columns=vectorizer.get_feature_names_out())
df.head()

able  absolutely  accommodate  actually  add  addition  afternoon  ago  \
0   0.0         0.0          0.0  0.000000  0.0       0.0        0.0  0.0   
1   0.0         0.0          0.0  0.000000  0.0       0.0        0.0  0.0   
2   0.0         0.0          0.0  0.000000  0.0       0.0        0.0  0.0   
3   0.0         0.0          0.0  0.129537  0.0       0.0        0.0  0.0   
4   0.0         0.0          0.0  0.000000  0.0       0.0        0.0  0.0   

   ahead   al  ...  write     wrong      year      yelp       yes  yesterday  \
0    0.0  0.0  ...    0.0  0.000000  0.000000  0.000000  0.000000        0.0   
1    0.0  0.0  ...    0.0  0.000000  0.128345  0.000000  0.171716        0.0   
2    0.0  0.0  ...    0.0  0.000000  0.000000  0.323791  0.000000        0.0   
3    0.0  0.0  ...    0.0  0.124057  0.000000  0.000000  0.000000        0.0   
4    0.0  0.0  ...    0.0  0.000000  0.000000  0.000000  0.000000        0.0   

   young  yum  yummy  zero  
0    0.0  0.0    0.0   0.0  
1    0.0  0.0    0.0   0.0  
2    0.0  0.0    0.0   0.0  
3    0.0  0.0    0.0   0.0  
4    0.0  0.0    0.0   0.0  

[5 rows x 800 columns]

In [53]:
from sklearn.linear_model import LogisticRegression


In [54]:
LR = LogisticRegression()
LR.fit(X_train_vect, Y_train)


/Users/vijaykumardevalla/anaconda3/lib/python3.11/site-packages/sklearn/linear_model/_logistic.py:460: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression()

In [55]:
predicted_lr = LR.predict(X_test_vect)
performance_lr = metrics.classification_report(Y_test, predicted_lr, target_names=['0', '1'])
print(performance_lr)


              precision    recall  f1-score   support

           0       0.86      0.81      0.84      2938
           1       0.92      0.94      0.93      6692

    accuracy                           0.90      9630
   macro avg       0.89      0.88      0.88      9630
weighted avg       0.90      0.90      0.90      9630



# Conclusion for lexicon based sentiment analysis

In [56]:
v_predicted = []
for text in X_test:
  sent= sentiment.polarity_scores(text)
  if sent['compound']>0:
    v_predicted.append(1)
  else:
    v_predicted.append(0)
v_performance = metrics.classification_report(Y_test,v_predicted, target_names= ['0', '1'])
print(v_performance)

              precision    recall  f1-score   support

           0       0.89      0.39      0.55      2938
           1       0.79      0.98      0.87      6692

    accuracy                           0.80      9630
   macro avg       0.84      0.69      0.71      9630
weighted avg       0.82      0.80      0.77      9630



Observations:

Model Performance: All three machine learning models demonstrate good performance with accuracy above 0.88. SVC performs the best in terms of accuracy (0.91), followed closely by Logistic Regression (0.90) and Naive Bayes (0.88).

Class Imbalance Handling: The class distribution (2938 for class 0 and 6692 for class 1) suggests a class imbalance. Despite this, all models perform relatively well for both classes, with slightly better performance on the majority class (class 1). This is a typical scenario in sentiment analysis tasks.

Precision and Recall: SVC and Logistic Regression have similar precision and recall scores, which are marginally better than those of Naive Bayes. This suggests they might be better at minimizing both false positives and false negatives.

F1-Score: The F1-score, which balances precision and recall, is consistently higher for class 1 across all models, indicating better performance on the positive class.


While lexicon-based sentiment analysis provided a simple and quick method for sentiment classification, machine learning models, as demonstrated by the provided outputs, tend to offer more robust and context-aware sentiment analysis. They yielded higher accuracy and are more adaptable to different types of text and domains. Among the machine learning models, SVC and Logistic Regression show slightly better overall performance compared to Naive Bayes in this particular task, particularly in handling both the positive and negative classes effectively.






#Input 2

In [62]:
text= ['The service is good, but location is hard to find. Sanitation is not very good with old facilities. Food served tasted extremely fishy, making us difficult to even finish it.'
,'The restaurant is definitely one of my favorites and of my family as well. I was especially impressed with my visit a few days ago. The place is clean, and you just need to wait for fewer than 10 minutes to get food served. And of course, the food is absolutely delicious!'
, 'I appreciated the friendly staff. The food was good, not amazing. The service was not prompt but almost acceptable. A reliable spot for a regular meal, but nothing extraordinary.']

In [64]:
dataframe = pd.DataFrame(text, columns=['text'])

In [67]:
dataframe.head()

text  \
0  The service is good, but location is hard to f...   
1  The restaurant is definitely one of my favorit...   
2  I appreciated the friendly staff. The food was...   

                                      text_processed  
0  service good , location hard find . sanitation...  
1  restaurant definitely favorite family . especi...  
2  appreciate friendly staff . food good , amazin...

In [66]:
dataframe['text_processed'] = dataframe['text'].apply(normalize, lowercase=True, remove_stopwords=True)

In [68]:
X3_vect = cv.transform(dataframe['text_processed'])

In [69]:
predicted_sentiment = LR.predict(X3_vect)


In [70]:
dataframe['predicted_sentiment'] = predicted_sentiment


In [71]:
dataframe.head()

text  \
0  The service is good, but location is hard to f...   
1  The restaurant is definitely one of my favorit...   
2  I appreciated the friendly staff. The food was...   

                                      text_processed  predicted_sentiment  
0  service good , location hard find . sanitation...                    1  
1  restaurant definitely favorite family . especi...                    1  
2  appreciate friendly staff . food good , amazin...                    1

In [73]:
# Predict sentiment probabilities
probabilities = LR.predict_proba(X3_vect)

# Define thresholds
positive_threshold = 0.7
negative_threshold = 0.3

# Classify each review
classifications = []
for prob in probabilities:
    if prob[1] >= positive_threshold:
        classifications.append('positive')
    elif prob[1] <= negative_threshold:
        classifications.append('negative')
    else:
        classifications.append('neutral')

# Add classifications to the DataFrame
dataframe['classification'] = classifications

# Display the DataFrame with classification and probabilities
dataframe['positive_probability'] = probabilities[:, 1]
print(dataframe)


                                                text  \
0  The service is good, but location is hard to f...   
1  The restaurant is definitely one of my favorit...   
2  I appreciated the friendly staff. The food was...   

                                      text_processed  predicted_sentiment  \
0  service good , location hard find . sanitation...                    1   
1  restaurant definitely favorite family . especi...                    1   
2  appreciate friendly staff . food good , amazin...                    1   

  classification  positive_probability  
0        neutral              0.586172  
1       positive              0.979902  
2       positive              0.982168  


Review 0: Classified as 'neutral' with a 58.6% positive sentiment probability. This makes sense because the review has mixed sentiments – it mentions good service but also points out negative aspects like location and sanitation. The model's probability suggests a slight positive bias but not strong enough to fully outweigh the negatives.

Review 1: Labeled as 'positive', and the model is 98% sure about this. This aligns well with the review's content, which is full of praises for the restaurant. The high probability reflects the clearly positive tone of the review.

Review 2: Also marked as 'positive' with a high confidence of 98.2%. While the review does include a minor critique, it primarily expresses satisfaction, particularly with the staff and food quality. The model seems to recognize the overall positive sentiment despite the slight criticism.



In [74]:
#Acknowledgement

I have taken help from chatgpt to finish this assignment. Although I have taken help, I know what code I am using and why I am using.